### PR TITLE
dev: improved handling of pending transactions in database

### PR DESCRIPTION
**Pull request by tcoratger**
Time spent on this PR: 2h

Resolves: #916

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

- During the sending of raw transactions, after signature recovering, a pending transaction database is populated before firing a transaction to Starknet for processing. This database collection aims to be able to query the database based on the hash of the transaction before receiving the receipt.
- The recovery of the chain ID with the handling of the error is moved to the start of the function because it is an inexpensive operation compared to the decoding of the transaction. It is therefore better to execute it first before decoding the transaction possibly for nothing if the chain ID is not correct.

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: tests/eth_provider.rs [hunk 1]
```diff
@@ -402,3 +405,91 @@ async fn test_to_starknet_block_id(#[future] katana: Katana, _setup: ()) {
...
+        value: U256::from(1000),
+        input: Bytes::default(),
+        max_fee_per_gas: 875000000,
+        max_priority_fee_per_gas: 0,
+        access_list: Default::default(),
+    });
+
+    // Sign the transaction
+    let signature = sign_message(
+        B256::from_str("0x02bbf4f9fd0bbb2e60b0316c1fe0b76cf7a4d0198bd493ced9b8df2a3a24d68a")
```
Comment (greged93) [inline tests/eth_provider.rs:431]: You can replace this with `katana.eoa().private_key`, just to avoid a hardcoded value

**Review by greged93 — COMMENTED** [review]
just a nit. Will you follow up with a PR in order to add a view in order to retrieve these pending txs and merge them with the indexed txs?

File: tests/eth_provider.rs (hunk 1 above)
Comment (tcoratger) [reply tests/eth_provider.rs]: Oh nice didn't know it exists, just fixed

**Comment by tcoratger** [comment]
> just a nit. Will you follow up with a PR in order to add a view in order to retrieve these pending txs and merge them with the indexed txs?

Yes I thought about a follow up PR dedicated to the view via get transaction by hash, this one being dedicated to database population.

**Review by greged93 — APPROVED** [review]

**Pull request merged**